Adjust nfs4_pwrite and nfs4_pread behavior to be more like C

scottnuma opened the pull request:
- Implement `nfs4_pwrite` to return bytes written, else -1 upon error
- `struct nfs4` modified to hold the `error_num` in addition to `error_string`, as the error number was previously given thru `pwrite`'s return value
- `sqllite_nfs4/sqllite_nfs4.c` has yet to be compile checked

scottnuma committed:
Return number bytes read/write

scottnuma committed:
Modify usage of nfs4_pwrite to reflecct C's write

- struct nfs4 now has error_num field. error_num was previously given
only from the return of status. error_num is used however, in files like
sqllite_nfs4/sqllite_nfs4.c.
- write may not write entire file without error. multiple writes may be
necessary

scottnuma commented on test/sfs.py:
```diff
@@ -117,12 +117,11 @@ def read(self, size=-1):
-                raise io.UnsupportedOperation("not writable") 
+        bytes_written = c_helper.nfs4_pwrite(self._file, content_bytes, self._pos, len(content_bytes))
+        if bytes_written < 0:
+            # if write_status == NFS4ERR_OPENMODE:
```
I still need to dig out the error_num from file->nfs4->err_num

jssmith replied:
Ok, 'looks like we're getting close.

scottnuma committed:
Adjust usage of nfs4_pread to match C style

- Modify looped pread and pwrite to change offset and length

scottnuma committed:
Update Python usage of nfs4_pread

jssmith commented on nfsv4/test/shell.c:
```diff
@@ -258,7 +258,17 @@ static value read_command(client c, vector args)
+
+    bytes remaining = n.size;
+    bytes total_read = 0;
+    while (total_read < n.size) {
```
Might be good to have a helper function here to read fully since this loop appears more than once in the shell.

jssmith commented on sqllite_nfs4/sqllite_nfs4.c:
```diff
@@ -58,7 +58,27 @@ static int nfs4Read(sqlite3_file *pFile,
+            if (f->ad->trace) {
+                eprintf ("status %s\n", nfs4_error_string(f->ad->c));
+            }
+            switch(f->f->c->error_num) {
```
can we say something like `translate_status(f->ad, f_f_c_error_num)`? to avoid duplicating the decoding switch statement.

jssmith commented on sqllite_nfs4/sqllite_nfs4.c:
```diff
@@ -69,7 +89,28 @@ static int nfs4Write(sqlite3_file *pFile,
+            if (f->ad->trace) {
+                eprintf ("status %s\n", nfs4_error_string(f->ad->c));
+            }
+            switch(f->f->c->error_num) {
```
As above, try to use `translate_status`

jssmith commented on nfsv4/nfs4_internal.h:
```diff
@@ -47,6 +47,7 @@ struct nfs4 {
     u32 maxreqs;
     buffer hostname;
     vector outstanding; // should be a map
+    int error_num;
```
maybe call this `nfs_error_num`

jssmith reviewed (commented):
This looks pretty good. Please address these few comments.

scottnuma committed:
Condense sending RPCs in pread and pwrite

scottnuma committed:
DRY for complete read and write in shell

scottnuma committed:
Update the direction for using shell

scottnuma committed:
Fix full_read_write return check

scottnuma committed:
DRY async_write_command with full_read_write

scottnuma committed:
DRY translate_status

jssmith reviewed (commented):
These changes look good and I think we're close.

What are you doing for testing?

scottnuma commented:
I'm still going through testing process,

`nfsv4_read_write_test` is still working as previous - this checks `pread` and `pwrite`, though I haven't figured out yet how to create an error for testing.

I was able to test the `read` command for the shell, but not the `write`. The `open write /file.txt` command segfaults both before and after the changes, which I think is consistent with the state we left the shell at.

Also still yet to compile or test the `ssqlite_nfs4` changes.

scottnuma committed:
Fold looping read and writes into single call

scottnuma committed:
Fix shell operations and add documentation

- Change pop_path to use next arguments rather than last
- Fix generate's allocate_buffer
- Print the content when printing a buffer, such as w/ READ

scottnuma committed:
Add to shell documentation

scottnuma committed:
Merge branch 'shell' into read_write_return

scottnuma committed:
Add skipped merge conflict changes

scottnuma committed:
Correctly decide between rpc_send and transact

scottnuma committed:
Choose non-zero seed for generate

- Ensures that the first character is not \0

scottnuma commented:
After some fixes to the shell, I've tested on both `read` and `write` and fixed up the documentation of how shell works.

I've now tested with sizes that require `nfs4_pread` to use multiple `rpc_send` and `transact`.
